File: leetcode/Dynamic Programming/house_robber.c
// https://leetcode.com/problems/house-robber/

#define MAX(a,b) (a) > (b) ? (a) : (b)

int rob(int* nums, int numsSize) {
    int* res = malloc(sizeof(int) * (numsSize + 2));

    memcpy(res, nums, sizeof(int) * numsSize);
    res[numsSize] = 0;
    res[numsSize+1] = 0;

    for (int i = 2; i < numsSize + 2; i++) {
        if (i == 2) {
            res[i] += res[i-2];
            continue;
        }
        res[i] += MAX(res[i-2], res[i-3]);
    }

    return res[numsSize+1];
}

File: leetcode/Dynamic Programming/house_robber.erl
%% https://leetcode.com/problems/house-robber/

-module(house_robber_II).

-compile(export_all).

-spec rob(Nums :: [integer()]) -> integer().
rob(Nums) ->
  rob(Nums, []).

rob([], [V1 | [V2 | _]]) ->
    max(V1, V2);
rob([], [V1 | _]) ->
    V1;
rob(_Nums = [H | T], Arr) ->
    NextVal = case Arr of
        [] -> H; %% empty arr
        [_ | [V1 | [V2 | _]]] -> H + max(V1, V2);  %% >3 elements in arr
        [_ | [V1 | _]] -> H + V1;  %% 2 elements in arr
        [_ | _] -> H  %% 1 element in arr
    end,
    rob(T, [NextVal | Arr]).

